File: src/main/java/org/uno/engine/Game.java
/*
 * Copyright (C) 2020  Fábio Furtado
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.uno.engine;

import org.uno.engine.objects.*;
import org.uno.enums.CardColour;
import org.uno.enums.CardType;
import org.uno.enums.SpecialCardSymbol;
import org.uno.enums.WildCardSymbol;
import org.uno.exceptions.engineExceptions.CardIndexOutOfHandBoundsException;
import org.uno.exceptions.engineExceptions.EngineException;
import org.uno.exceptions.engineExceptions.InvalidOptionException;
import org.uno.exceptions.engineExceptions.MissingColourForWildCardException;
import org.uno.util.Stack;
import org.uno.util.Vector;

import java.util.Arrays;
import java.util.Random;


/**
 * @see UnoGame
 *
 * @author Fábio Furtado
 */
public final class Game implements UnoGame {

    /**
     * All the cards on the deck. The deck is where the players draw new cards.
     * It will get refilled with the cards from the table if it's close to get
     * empty.
     */
    private final Stack<Card> deck;

    /**
     * The collection of the cards which have been played
     */
    private final Stack<Card> table;

    /**
     * The players of this game
     */
    private final Player[] players;

    /**
     * The index of the player in turn
     */
    private int turn;

    /**
     * The index of the player who played previously
     */
    private int previous;

    /**
     * The direction to which the turn to play moves. 1 means it's moving towards
     * the player with higher index, -1 means it's moving towards the one with
     * lower index.
     */
    private int direction;

    /**
     * The number of cards each player starts with
     */
    private static final int N_OF_STARTING_CARDS = 7;

    /**
     * The winner of this game. It's meant to be null until the game is over
     */
    private Player winner;

    /**
     * Creates a new instance.
     *
     * @param players players for this game
     */
    Game(Player[] players) {
        this.deck = new DeckGenerator().next();
        this.table = new Stack<>();
        this.players = new Player[players.length];
        System.arraycopy(players, 0, this.players, 0, players.length);
        this.turn = new Random().nextInt(players.length);
        this.direction = 1;
        this.previous = turn;
        this.winner = null;
        distributeAndFlip();
    }

    /**
     * @see UnoGame#getNumberOfPlayers()
     */
    @Override
    public int getNumberOfPlayers() {
        return players.length;
    }

    /**
     * @see UnoGame#getPlayer(int)
     */
    @Override
    public Player getPlayer(int index) {
        return players[index].clone();
    }

    /**
     * @see UnoGame#getPlayer(String)
     */
    @Override
    public Player getPlayer(String id) {
        for (int i = 0; i < players.length; i++) {
            if (players[i].getId().equals(id))
                return players[i].clone();
        }
        return null;
    }

    /**
     * @see UnoGame#getPlayers()
     */
    @Override
    public Player[] getPlayers() {
        return Arrays.copyOf(players, players.length);
    }

    /**
     * @see UnoGame#getPreviousPlayer()
     */
    @Override
    public Player getPreviousPlayer() {
        return players[previous].clone();
    }

    /**
     * @see UnoGame#getIndex(String)
     */
    @Override
    public int getIndex(String id) {
        for (int i = 0; i < players.length; i++) {
            if (players[i].getId().equals(id))
                return i;
        }
        return -1;
    }

    /**
     * @see UnoGame#getPlayerInTurn()
     */
    @Override
    public Player getPlayerInTurn() {
        return players[turn].clone();
    }

    /**
     * @see UnoGame#getDeckTop()
     */
    @Override
    public Card getDeckTop() {
        return deck.peek().clone();
    }

    /**
     * @see UnoGame#getTableTop()
     */
    @Override
    public Card getTableTop() {
        return table.peek().clone();
    }

    /**
     * @see UnoGame#getWinner()
     */
    @Override
    public Player getWinner() {
        if (isOver())
            return winner.clone();
        throw new IllegalStateException("This game is not over yet!");
    }

    /**
     * @see UnoGame#isOver()
     */
    @Override
    public boolean isOver() {
        return winner != null;
    }

    /**
     * Distributes the cards to the players and puts the first
     * card on the table.
     */
    private void distributeAndFlip() {
        for (Player player : players) {
            for (int i = 0; i < N_OF_STARTING_CARDS; i++)
                player.addToHand(deck.pop());
        }

        // Making sure the first card to be flipped is a numeric one
        Stack<Card> temp = new Stack<>();
        while (deck.peek().getType() != CardType.NUMERIC)
            temp.push(deck.pop());
        table.push(deck.pop());
        for (int i = 0; i < temp.size(); i++)
            deck.push(temp.pop());
    }

    /**
     * @see UnoGame#goBot()
     */
    @Override
    public GameCommand goBot() {
        GameCommand command = null;
        if (getPlayerInTurn().getClass() == BotPlayer.class) {
            Bot bot = (BotPlayer) getPlayerInTurn();
            command = bot.makeMove(this);
            try {
                executeMove(command);
            } catch (EngineException e) {
                // Bot should not cause game exceptions
                e.printStackTrace();
            }
        }
        return command;
    }

    /**
     * @see UnoGame#executeMove(GameCommand)
     */
    @Override
    public int executeMove(GameCommand command)
            throws InvalidOptionException,
            CardIndexOutOfHandBoundsException,
            MissingColourForWildCardException {
        if (!isOver()) {
            makeSureDeckDoesNotGetEmpty();
            int returnValue = 1;
            checkCommandValidity(command);

            if (command.getOption() == 0) {
                draw();
                returnValue = 0;
            } else if (command.getOption() == 1) {
                returnValue = play(command);
                if (players[previous].getHand().isEmpty())
                    winner = players[previous];
            }
            return returnValue;
        } else
            throw new IllegalStateException("This game is already over!");
    }

    private void draw() {
        players[turn].addToHand(deck.pop());
        previous = turn;
        moveToNextPlayer();
    }

    private int play(GameCommand gameCommand) {
        int returnValue = 1;
        Card card = players[turn].getHand().get(gameCommand.getIndex());

        if (isCardValid(card)) {
            previous = turn;
            table.push(players[turn].takeFromHand(gameCommand.getIndex()));

            // execute the action demanded by the played card
            switch (table.peek().getType()) {
            case SPECIAL:
                doCaseSpecial();
                break;

            case WILD:
                doCaseWild(gameCommand);
                break;

            case NUMERIC:

                // No special interactions to do in case of an numeral card so
                // we just have to move to the next player
                moveToNextPlayer();
                break;
            }
            returnValue = 0;
        }
        return returnValue;
    }

    /**
     * Cheks if the given command is valid. Throws exceptions in case problems are
     * found
     * @param command command to be checked
     * @throws MissingColourForWildCardException if the command plays a wild card
     *                                           but does not have a chosen colour
     * @throws CardIndexOutOfHandBoundsException if the command plays a card out
     *                                           of the player in turn hand's
     *                                           index bounds
     * @throws InvalidOptionException            If the given option is not valid
     *
     * @throws IllegalStateException             if an index and/or colour were given
     *                                           when the option is to draw
     */
    private void checkCommandValidity(GameCommand command)
            throws MissingColourForWildCardException,
            CardIndexOutOfHandBoundsException,
            InvalidOptionException {
        if (command.getOption() == 0) {
            if (command.getIndex() != -1 || command.getColour() != null) {
                throw new IllegalStateException(
                        "Not expecting values for index and/or colour" +
                                                          "when option is 0"
                );
            }
        } else if (command.getOption() == 1) {
            if (command.getIndex() > players[turn].getHand().length() - 1 ||
                    command.getIndex() < 0)
                throw new CardIndexOutOfHandBoundsException(String.format(
                        "%d is out of hand range of %d",
                        command.getIndex(), players[turn].getHand().length())
                );
            if (players[turn].getHand().get(command.getIndex()).getClass() ==
                    WildCard.class && command.getColour() == null) {
                throw new MissingColourForWildCardException();
            }
        } else
            throw new InvalidOptionException("Invalid option");
    }

    private void doCaseSpecial() {
        SpecialCard tableTop = (SpecialCard) table.peek();
        if (tableTop.getSymbol() == SpecialCardSymbol.DRAW_2) {
            moveToNextPlayer();
            for (int i = 0; i < 2; i++)
                players[turn].addToHand(deck.pop());
            moveToNextPlayer();
        } else if (tableTop.getSymbol() == SpecialCardSymbol.REVERSE) {
            revert();
            if (players.length > 2)
                moveToNextPlayer();
        } else { // skip card
            jumpAPlayer();
        }
    }

    /**
     * Sets the chosen colour for the wild card and adds 4 cards to the next
     * player's hand in case the card is a wild 4.
     *
     * @param command command of the move
     */
    private void doCaseWild(GameCommand command) {
        CardColour colour = command.getColour();

        WildCard c = (WildCard) table.pop();
        c.setPickedColour(colour);
        table.push(c);

        WildCard tableTop = (WildCard) table.peek();
        if (tableTop.getSymbol() == WildCardSymbol.DRAW_4) {
            moveToNextPlayer();
            for (int i = 0; i < 4; i++)
                players[turn].addToHand(deck.pop());
            moveToNextPlayer();
        } else {
            moveToNextPlayer();
        }
    }

    /**
     * If the deck size is under 4, the cards from the table will be reshuffled
     * and passed the the deck so it doesn't get empty.
     */
    private void makeSureDeckDoesNotGetEmpty() {
        if (deck.size() < 4) {
            Card tableTop = table.pop();
            Vector<Card> temp = new Vector<>(table.size());
            for (int i = 0; i < table.size(); i++)
                temp.add(i, table.pop());
            table.push(tableTop);
            temp.shuffle();
            for (int i = 0; i < temp.length(); i++)
                deck.push(temp.get(i));
        }
    }

    /**
     * @see UnoGame#isCardValid(Card)
     */
    @Override
    public boolean isCardValid(Card card) {
        CardType tableTopType = table.peek().getType();
        boolean isValid;

        if (tableTopType == CardType.NUMERIC) {
            isValid = checkValidityForNumericTop(card);
        } else if (tableTopType == CardType.SPECIAL) {
            isValid = checkValidityForSpecialTop(card);
        } else { // wild in table
            isValid = checkValidityForWildTop(card);
        }
        return isValid;
    }

    private boolean checkValidityForNumericTop(Card card) {
        boolean returnValue;
        NumericCard tableTop = (NumericCard) table.peek();

        if (card.getType() == CardType.SPECIAL || card.getType() == CardType.NUMERIC) {
            if (card.getType() == CardType.SPECIAL) {
                SpecialCard numericCard = (SpecialCard) card;
                returnValue = numericCard.getColour() == tableTop.getColour();
            } else {
                NumericCard numericCard = (NumericCard) card;
                returnValue = numericCard.getColour() == tableTop.getColour()
                        || numericCard.getNumber() == tableTop.getNumber();
            }
        } else { // wild card
            returnValue = true;
        }
        return returnValue;
    }

    private boolean checkValidityForSpecialTop(Card card) {
        SpecialCard tableTop = (SpecialCard) table.peek();
        boolean returnValue;

        if (card.getType() == CardType.NUMERIC) {
            NumericCard specialCard = (NumericCard) card;
            returnValue = specialCard.getColour() == tableTop.getColour();
        } else if (card.getType() == CardType.SPECIAL) {
            SpecialCard specialCard = (SpecialCard) card;
            returnValue = specialCard.getColour() == tableTop.getColour()
                    || specialCard.getSymbol() == tableTop.getSymbol();
        } else {

            // No restrictions for wild card
            returnValue = true;
        }
        return returnValue;
    }

    private boolean checkValidityForWildTop(Card card) {
        WildCard tableTop = (WildCard) table.peek();
        boolean returnValue = false;

        if (card.getType() == CardType.NUMERIC) {
            NumericCard numericCard = (NumericCard) card;
            returnValue = numericCard.getColour() == tableTop.getPickedColour();
        } else if (card.getType() == CardType.SPECIAL) {
            SpecialCard specialCard = (SpecialCard) card;
            returnValue = specialCard.getColour() == tableTop.getPickedColour();
        } else if (card.getType() == CardType.WILD) {
            returnValue = true;
        }
        return returnValue;
    }

    /**
     * Moves the turn to the next player
     */
    private void moveToNextPlayer() {
        if (direction > 0 && turn == players.length - 1)
            turn = 0;
        else if (direction < 0 && turn == 0)
            turn = players.length - 1;
        else
            turn += direction;
    }

    /**
     * Jumps a player
     */
    private void jumpAPlayer() {
        moveToNextPlayer();
        moveToNextPlayer();
    }

    /**
     * Reverts the game. Usually done when someone plays a revert card
     */
    private void revert() {
        direction *= -1;
    }
}

File: src/main/java/org/uno/engine/Game.kt
/*
 * Copyright (C) 2020  Fábio Furtado
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.uno.engine

import org.uno.engine.engineExceptions.CardIndexOutOfHandBoundsException
import org.uno.engine.engineExceptions.GameRulesException
import org.uno.engine.engineExceptions.MissingColourForWildCardException
import org.uno.engine.objects.*
import org.uno.util.Stack
import kotlin.random.Random

/**
 * @see UnoGame
 *
 * @author Fábio Furtado
 */
class Game private constructor(_players: Array<Player>, _deck: Stack<Card>,
                               _table: Stack<Card>,_turn: Int, _previous: Int,
                               _direction: Int, _winner: Player?) : UnoGame {

    /**
     * All the cards on the deck. The deck is where the players draw new cards.
     * It will get refilled with the cards from the table if it's close to get
     * empty.
     */
    private val deck = _deck

    /**
     * Cards which have been played
     */
    private val table = _table

    /**
     * The index of the player in turn
     */
    private var turn = _turn

    /**
     * The index of the player who played previously
     */
    private var previous = _previous

    /**
     * The direction to which the turn to play moves. 1 means it's moving towards
     * the player with higher index, -1 means it's moving towards the one with
     * lower index.
     */
    private var direction = _direction

    /**
     * The number of cards each player starts with
     */
    private val startingCards = 7

    /**
     * The players of this game
     */
    private val players = _players

    override var winner: Player? = _winner
        private set
        get() = field?.clone()

    /**
     * @see UnoGame.playerInTurn
     */
    override val playerInTurn
        get() = players[turn].clone()

    /**
     * @see UnoGame.previousPlayer
     */
    override val previousPlayer
        get() = players[previous].clone()

    override val deckTop: Card
        get() = deck.peek().clone()

    override val tableTop: Card
        get() = table.peek().clone()

    override val numberOfPlayers: Int
        get() = players.size

    override val isOver: Boolean
        get() = winner != null
    /**
     * Distributes the cards to the players and puts the first
     * card on the table.
     */
    private fun distributeAndFlip() {
        for (player in players) {
            var i = 0
            while (i < startingCards) {
                player.addToHand(deck.pop())
                i++
            }
        }

        // Making sure the first card to be flipped is a numeric one
        val aux = Stack<Card>()
        while (deck.peek().type != CardType.NUMERIC)
            aux.push(deck.pop())
        table.push(deck.pop())
        for (card in aux)
            deck.push(card)
    }

    /**
     * @see UnoGame#getPlayer(int)
     */
    override fun getPlayer(index: Int): Player = players[index].clone()

    /**
     * @see UnoGame.getPlayer
     */
    override fun getPlayer(id: String?): Player? {
        for (player in players) {
            if (id.equals(player.id))
                return player.clone()
        }
        return null
    }

    /**
     * @see UnoGame.getIndex
     */
    override fun getIndex(id: String): Int {
        for (index in players.indices) {
            if (players[index].id == id)
                return index
        }
        return -1
    }

    /**
     * @see UnoGame#isCardValid(Card)
     */
    override fun isCardValid(card: Card): Boolean {
        return when (card.type) {
            CardType.NUMERIC -> checkValidityForNumeric(card as NumericCard)
            CardType.SPECIAL -> checkValidityForSpecial(card as SpecialCard)

            // Wild cards are always valid
            else -> true
        }
    }

    private fun checkValidityForNumeric(card: NumericCard): Boolean {
        val top = table.peek()
        var isValid = false
        if (top is Numeric && card.number == (top as Numeric).number)
            isValid = true
        if (top is Colourful && card.colour === (top as Colourful).colour)
            isValid = true
        if (top is Wild && card.colour === (top as Wild).pickedColour)
            isValid = true
        return isValid
    }

    private fun checkValidityForSpecial(card: SpecialCard): Boolean {
        val top = table.peek()
        var isValid = false
        if (top is Colourful && card.colour === (top as Colourful).colour)
            isValid = true
        if (top is Symbolic && card.symbol === (top as Symbolic).symbol)
            isValid = true
        if (top is Wild && card.colour === (top as Wild).pickedColour)
            isValid = true
        return isValid
    }

    /**
     * @see UnoGame.executeMove
     */
    @Throws(CardIndexOutOfHandBoundsException::class,
            MissingColourForWildCardException::class)
    override fun executeMove(command: GameCommand): Int {
        return if (!isOver) {
            checkMoveValidity(command)
            makeSureDeckDoesNotGetEmpty()
            var returnValue = 1
            if (command.option == 0) {
                draw()
                returnValue = 0
            } else if (command.option == 1) {
                returnValue = play(command)
                if (previousPlayer.hand.isEmpty())
                    winner = players[turn]
            }
            returnValue
        } else throw IllegalStateException("This game is already over!")
    }

    /**
     * Checks if the given {@link GameCommand} is valid. Throws exceptions
     * if problems are found.
     *
     * @throws MissingColourForWildCardException if the command plays a wild card
     *                                           but does not have a chosen colour
     * @throws CardIndexOutOfHandBoundsException if the command plays a card out
     *                                           of the player in turn hand's
     *                                           index bounds
     */
    @Throws(MissingColourForWildCardException::class,
            CardIndexOutOfHandBoundsException::class)
    private fun checkMoveValidity(command: GameCommand) {
        if (command.option == 1) {
            if (command.index > players[turn].hand.size - 1)
                throw CardIndexOutOfHandBoundsException(
                        "${command.index} is out of hand range of ${players[turn].hand.size}"
                )
            else if (players[turn].hand[command.index] is Wild && command.colour == null)
                throw MissingColourForWildCardException()
        }
    }

    /**
     * If the deck size is lesser than 5, the cards from the table will be
     * reshuffled and passed the the deck so it doesn't get empty.
     */
    private fun makeSureDeckDoesNotGetEmpty() {
        if (deck.size() < 5) {
            val lastInDeck = Stack<Card>()
            val tableList = ArrayList<Card>()
            val tableTop = table.pop()
            while (!deck.isEmpty)
                lastInDeck.push(deck.pop())
            while (!table.isEmpty)
                tableList.add(table.pop())
            tableList.shuffle()
            for (card in tableList)
                deck.push(card)
            for (card in lastInDeck)
                deck.push(card)
            table.push(tableTop)
        }
    }

    private fun draw() {
        players[turn].addToHand(deck.pop())
        updatePrevious()
        move()
    }

    private fun play(command: GameCommand): Int {
        val card = players[turn].hand[command.index]
        return if (isCardValid(card)) {
            updatePrevious()
            when (card.type) {
                CardType.SPECIAL -> playSpecial(command)
                CardType.WILD -> playWild(command)

                // Numeric card
                else -> playNumeric(command)
            }
            0
        }
        else 1
    }

    private fun playSpecial(command: GameCommand) {
        table.push(players[turn].takeFromHand(command.index))
        when ((table.peek() as Symbolic).symbol) {
            SpecialCardSymbol.DRAW_2 -> {
                move()
                for (i in 0..1) players[turn].addToHand(deck.pop())
                move()
            }
            SpecialCardSymbol.REVERSE -> {
                revert()
                if (players.size > 2) move()
            }

            // Skip card
            else -> move(2)
        }
    }

    /**
     * Sets the chosen colour for the wild card and adds 4 cards to the next
     * player's hand in case the card is a wild 4.
     *
     * @param command command of the move
     */
    private fun playWild(command: GameCommand) {
        (players[turn].hand[command.index] as Wild).pickedColour = command.colour
        table.push(players[turn].takeFromHand(command.index))

        when ((table.peek() as Symbolic).symbol) {
            WildCardSymbol.DRAW_4 -> {
                move()
                for (i in 0 until 4)
                    players[turn].addToHand(deck.pop())
                move()
            }
            else -> move()
        }
    }

    private fun playNumeric(command: GameCommand) {
        table.push(players[turn].takeFromHand(command.index))
        move()
    }

    /**
     * @see UnoGame#goBot()
     */
    override fun goBot(): GameCommand {
        if (players[turn] is Bot) {
            val move = (players[turn] as Bot).makeMove(this)
            executeMove(move)
            return move
        }
        throw IllegalStateException()
    }

    /**
     * Moves the turn as many times as requested.
     *
     * @param repeat number of times to repeat this operation
     */
    private fun move(repeat: Int) {
        for (i in 0 until repeat) {
            if (direction == 1 && turn == players.size - 1)
                turn = 0
            else if (direction == -1 && turn == 0)
                turn = players.size - 1
            else turn += direction
        }
    }

    /**
     * Moves the turn to the next player.
     */
    private fun move() = move(1)

    /**
     * Updates the previous and previousPlayer properties. Should be called
     * whenever a move is about to be executed so it updates those values based
     * on the current turn and player in turn.
     */
    private fun updatePrevious() {
        previous = turn
    }

    /**
     * Reverts the turn direction. Usually done when a revert card is played.
     */
    private fun revert() {
        direction *= -1
    }

    /**
     * @see UnoGame#clone()
     */
    override fun clone() = instance(
            players.copyOf(),
            deck.clone(),
            table.clone(),
            turn, previous, direction )

    /**
     * Creates instances of [Game] class.
     */
    companion object Factory {

        /**
         * All bots name's will start with this prefix.
         */
        private const val BOT_PLAYER_NAME_PREFIX = "Bot"

        /**
         * Minimal number of players for each game.
         */
        const val MIN_NUMBER_OF_PLAYERS = 2

        /**
        * Maximal number of players for each game.
        */
        const val MAX_NUMBER_OF_PLAYERS = 8

        /**
         * This is to be used only and in no other situation other than:
         * to create a game with a previous state or to create a clone.
         * Appliance in other situations may produce unexpected behaviour. This
         * shouldn't be used especially to create a Game in a initial state.
         *
         * Make sure the following is true:
         *```
         * code players != null && players.length > 0 && deck != null &&
         * table != null && deck.size() + table.size() == 108 && turn > 0
         * && turn < players.length && (direction == 1 || direction == -1)}
         * ```
         * @param players   players for this instance
         * @param deck      deck for this instance
         * @param table     table for this instance
         * @param turn      turn for this instance
         * @param previous  previous for this instance
         * @param direction direction for this instance (1 or -1)
         */
        fun instance(players: Array<Player>,
                     deck: Stack<Card>,
                     table: Stack<Card>,
                     turn: Int,
                     previous: Int,
                     direction: Int) : Game {

            var winner:Player? = null
            for (player in players) {
                if (player.hand.isEmpty()) winner = player
            }
            return Game(players, deck, table, turn, previous, direction, winner)
        }

        /**
         * Creates a new Game with [numberOfBots] defining how many bot player's
         * it'll have and [humanPlayersNames] defining the number of human players.
         *
         * The sum of those parameters can't be lesser than [MIN_NUMBER_OF_PLAYERS]
         * or greater than [MAX_NUMBER_OF_PLAYERS].
         *
         * @param numberOfBots      number of bot's to add
         * @param humanPlayersNames array with the name of each human player
         * @return a new `Game` instance
         */
        fun instance(numberOfBots: Int, vararg humanPlayersNames: String) : Game {
            val players = arrayOfNulls<Player>(numberOfBots + humanPlayersNames.size)
            checkIfNumberOfPlayersIsLegal(players.size)
            for (i in 0 until numberOfBots)
                players[i] = BotPlayer(BOT_PLAYER_NAME_PREFIX.plus(i + 1))
            var j = 0
            for (i in numberOfBots until players.size)
                players[i] = HumanPlayer(humanPlayersNames[j++])

            val game = Game(players as Array<Player>, DeckGenerator().next(),
                    Stack(), Random.nextInt(0, players.size), 0, 1, null)
            game.distributeAndFlip()
            return game
        }

        /**
         * Checks if the given number of players is legal, throwing an Exception case
         * not.
         *
         * @param numberOfPlayers number of players to test
         * @throws GameRulesException if the given number of players is illegal
         */
        @Throws(GameRulesException::class)
        private fun checkIfNumberOfPlayersIsLegal(numberOfPlayers: Int) {
            var exceptionMessage: String? = null
            if (numberOfPlayers > MAX_NUMBER_OF_PLAYERS)
                exceptionMessage = "$numberOfPlayers players exceed the " +
                        "players limit of $MAX_NUMBER_OF_PLAYERS"
            else if (numberOfPlayers < MIN_NUMBER_OF_PLAYERS)
                exceptionMessage = "$numberOfPlayers players is not enough to " +
                        "start a game, minimum of $MIN_NUMBER_OF_PLAYERS" +
                        " is needed"
            if (exceptionMessage != null)
                throw GameRulesException(exceptionMessage)
        }
    }
}